src/gConsent/query/query.ts

```diff
@@ -0,0 +1,262 @@
+   */
+  issuedWithin?: "P1D" | "P7D" | "P1M" | "P3M";
+  /**
+   * Interval (expressed using ISO 8601) during which the Credential was revoked.
+   */
+  revokedWithin?: "P1D" | "P7D" | "P1M" | "P3M";
+  /**
+   * Number of items per page.
+   */
+  pageSize?: number;
+  /**
+   * Target page (for result pagination).
+   */
+  page?: string;
+};
+
+const FILTER_ELEMENTS: Array<keyof CredentialFilter> = [
+  "fromAgent",
+  "issuedWithin",
+  "page",
+  "pageSize",
+  "purpose",
+  "resource",
+  "revokedWithin",
+  "status",
+  "toAgent",
+  "type",
+] as const;
+
+// The following ensures that, when we add a value to the FILTER_ELEMENTS array, TS complains
+// if there are missing cases when handling the value.
+type SupportedFilterElements =
+  typeof FILTER_ELEMENTS extends Array<infer E> ? E : never;
+
+// The type assertion is okay in the context of the type guard.
+const isSupportedFilterElement = (x: unknown): x is SupportedFilterElements =>
+  FILTER_ELEMENTS.includes(x as keyof CredentialFilter);
+
+const PAGING_RELS = ["first", "last", "prev", "next"] as const;
+
+export type CredentialResult = {
+  /**
+   * Page of Access Credentials matching the query.
+   */
+  items: DatasetWithId[];
+  /**
+   * First page of query results.
+   */
+  first?: CredentialFilter;
+  /**
+   * Previous page of query results.
+   */
+  prev?: CredentialFilter;
+  /**
+   * Next page of query results.
+   */
+  next?: CredentialFilter;
+  /**
+   * Last page of query results.
+   */
+  last?: CredentialFilter;
+};
+
+function toCredentialFilter(url: URL): CredentialFilter {
+  const filter: CredentialFilter = {};
+  url.searchParams.forEach((value, key) => {
+    if (isSupportedFilterElement(key)) {
+      Object.assign(filter, { [`${key}`]: value });
+    }
+  });
+  return filter;
+}
+
+function hasItems(x: unknown): x is { items: unknown[] } {
+  const candidate = x as { items: unknown };
+  return candidate.items !== undefined && Array.isArray(candidate.items);
+}
+
+function isObjectWithId(x: unknown): x is { id: string } {
+  const candidate = x as { id: string };
+  return typeof candidate.id === "string";
+}
+
+async function toCredentialResult(
+  response: Response,
+): Promise<CredentialResult> {
+  const result: CredentialResult = { items: [] };
+  const linkHeader = response.headers.get("Link");
+  if (linkHeader !== null) {
+    const parsedLinks = LinkHeader.parse(linkHeader);
+    PAGING_RELS.forEach((rel) => {
+      const link = parsedLinks.get("rel", rel);
+      if (link.length > 1) {
+        throw Error(
+          `Unexpected response, found more than one ${rel} Link headers.`,
+        );
+      }
+      if (link.length === 0) {
+        return;
+      }
+      result[rel] = toCredentialFilter(new URL(link[0].uri));
+    });
+  }
+  const body = await response.json();
+  if (!hasItems(body) || !body.items.every(isObjectWithId)) {
+    throw Error(
+      `Unexpected response, no items found in ${JSON.stringify(body)}`,
```

Comment:
Based on my previous comment, this would throw if `items` as not defined as an error but would not throw if it was empty although the error message looks like it would cover that condition.

Reply:
That is a mistake from me in the message phrasing, an empty array is acceptable.

Reply:
Changed in 87c5e87e8d16028db4f11d0dc54ae488704ad08b